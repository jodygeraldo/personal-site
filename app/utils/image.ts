type AcceptedImageOutput = 'png' | 'webp'

function generateSources(
  originalUrl: string,
  output: AcceptedImageOutput[],
  set: Array<{ width: number; height: number; media: string }>,
) {
  let url = `//images.weserv.nl/?url=https://jodygeraldo.com${originalUrl}`
  const sources: Array<{
    url: string
    type: string
    media: string
    width: number
    height: number
  }> = []

  set.forEach((image) => {
    const attributes = {
      media: image.media,
      width: image.width,
      height: image.height,
    }

    if (output.includes('webp')) {
      sources.push({
        url: `${url}&w=${image.width}&h=${image.height}&output=webp`,
        type: 'image/webp',
        ...attributes,
      })
    }

    if (output.includes('png')) {
      sources.push({
        url: `${url}&w=${image.width}&h=${image.height}&af`,
        type: 'image/png',
        ...attributes,
      })
    }
  })

  return sources
}

function generateFallbackImage(
  originalUrl: string,
  width: number,
  height: number,
) {
  const fallback = `//images.weserv.nl/?url=https://jodygeraldo.com${originalUrl}&w=${width}&h=${height}&af`

  return { url: fallback, width: width, height: height }
}

function generatePictureSource(
  originalUrl: string,
  width: number,
  height: number,
  output: AcceptedImageOutput[],
  set: Array<{ width: number; height: number; media: string }>,
) {
  const fallback = generateFallbackImage(originalUrl, width, height)
  const sources = generateSources(originalUrl, output, set)

  return { fallback, sources }
}

export { generatePictureSource }
