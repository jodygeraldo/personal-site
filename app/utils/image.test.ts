import { generatePictureSource } from './image'

const mock = {
  url: '/assets/images/test.png',
  width: 512,
  height: 288,
  output: ['png', 'webp'] as ['png', 'webp'],
  set: [
    { width: 384, height: 216, media: '(max-width: 639px)' },
    {
      width: 464,
      height: 261,
      media: '(min-width: 640px) and (max-width: 1023px)',
    },
    { width: 512, height: 288, media: '(min-width: 1024px)' },
  ],
}

test('should return object to build html picture source', () => {
  expect(
    generatePictureSource(
      mock.url,
      mock.width,
      mock.height,
      mock.output,
      mock.set,
    ),
  ).toMatchSnapshot()
})
