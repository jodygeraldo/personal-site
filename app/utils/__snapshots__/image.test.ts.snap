// Vitest Snapshot v1

exports[`should return object to build html picture source 1`] = `
{
  "fallback": {
    "height": 288,
    "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=512&h=288&af",
    "width": 512,
  },
  "sources": [
    {
      "height": 216,
      "media": "(max-width: 639px)",
      "type": "image/webp",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=384&h=216&output=webp",
      "width": 384,
    },
    {
      "height": 216,
      "media": "(max-width: 639px)",
      "type": "image/png",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=384&h=216&af",
      "width": 384,
    },
    {
      "height": 261,
      "media": "(min-width: 640px) and (max-width: 1023px)",
      "type": "image/webp",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=464&h=261&output=webp",
      "width": 464,
    },
    {
      "height": 261,
      "media": "(min-width: 640px) and (max-width: 1023px)",
      "type": "image/png",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=464&h=261&af",
      "width": 464,
    },
    {
      "height": 288,
      "media": "(min-width: 1024px)",
      "type": "image/webp",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=512&h=288&output=webp",
      "width": 512,
    },
    {
      "height": 288,
      "media": "(min-width: 1024px)",
      "type": "image/png",
      "url": "//images.weserv.nl/?url=https://jodygeraldo.com/assets/images/test.png&w=512&h=288&af",
      "width": 512,
    },
  ],
}
`;
